### PR TITLE
My solution to the airport challenge

### PR DESCRIPTION
**Pull request by ffgi-es**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,4 @@ CoverallsRakefile
 analysis/**/*
 .gitignore-analysis
 
+rough_feature_test
```
Comment (alicelieutier) [inline .gitignore]: You could totally have pushed this - it's nice to see drafts too :)

File: lib/airport.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+require_relative 'weather'
+
+class Airport
+  DEFAULT_CAPACITY = 20
+  attr_reader :name, :capacity
+
+  def initialize(name, opts = {})
```
Comment (alicelieutier) [inline lib/airport.rb]: Replace `opts` with `options`

File: lib/plane.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    raise InvalidStateError, "Not in the air" unless flying?
+    return false unless airport.accept self
+
+    @airport = airport
+    true
+  end
+
+  def take_off
+    raise InvalidStateError, "Can't take off in the air" if flying?
+    return false unless @airport.take_off_slot self
```
Comment (alicelieutier) [inline lib/plane.rb]: This is a small comment, but it is generally not best practice to have a method that has a side effect as a conditional. Making this into 2 steps would clarify what happens:
```rb
take_off_attempt_success = @airport.take_off_slot
return false if !take_off_attempt_success
@airport = nil
 true

```

The other pattern you can use here, that would probably be more idiomatic is using errors and try/catch block.

File: spec/airport_spec.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  let(:plane2) { double(:plane, flying?: false) }
+  let(:weather) { double(:weather, stormy?: false) }
+
+  def bad_weather
+    allow(weather).to receive(:stormy?).and_return(true)
+  end
+
+  subject { Airport.new("Paris", weather: weather) }
+
+  before(:each) { allow(subject).to receive(:weather_report).and_return("sunny") }
```
Comment (alicelieutier) [inline spec/airport_spec.rb]: You should never stub a method of the class under test. So `allow(subject)...` is always a red flag.
Instead, you should be using what you prepared above, something like:
`allow(weather).to receive(:stormy?).and_return(false)`

File: spec/airport_spec.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  end
+
+  describe "#new" do
+    it "should accept an optional argument for custom capacity of 30" do
+      airport = Airport.new("Barcelona", capacity: 30)
+      expect(airport.capacity).to eq 30
+    end
+  end
+
+  describe "#count" do
```
Comment (alicelieutier) [inline spec/airport_spec.rb]: Nice!

File: spec/weather_spec.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require 'weather'
+
+describe Weather do
+  describe "#stormy?" do
+    it "should return either true or false" do
+      expect(subject.stormy?).to eq(true).or eq(false)
```
Comment (alicelieutier) [inline spec/weather_spec.rb]: Instead of this, you may want to stub randomness and expect a specific result.

**Review by alicelieutier — COMMENTED** [review]
Good solution!
I especially like the feature tests, and the code is quite readable.
Here are a few comments:

File: spec/airport_spec.rb (hunk 4 above)
Comment (ffgi-es) [reply spec/airport_spec.rb]: Yeah, I can see how that could be problematic. I believe that was a hangover from before I separated the Weather out into its own class as that method was actually completely removed in the end.

**Pull request closed without merging**